Userland/Games/Solitaire/Game.cpp

```diff
@@ -482,6 +498,76 @@ void Game::paint_event(GUI::PaintEvent& event)
     }
 }
 
+void Game::undo_save_move(CardStack& from, CardStack& to, NonnullRefPtrVector<Card> moved_cards)
+{
+    m_last_move.type = LastMove::Type::MoveCards;
+    m_last_move.from = &from;
+    m_last_move.cards = moved_cards;
+    m_last_move.to = &to;
+    if (on_undo_available)
+        on_undo_available(true);
+}
+
+void Game::undo_save_flip(Card& card)
+{
+    NonnullRefPtrVector<Card> cards;
+    cards.append(card);
+    m_last_move.type = LastMove::Type::FlipCard;
+    m_last_move.cards = cards;
+    if (on_undo_available)
+        on_undo_available(true);
+}
+
+void Game::perform_undo()
+{
+    if (!m_last_move.from)
+        return;
+
+    if (m_last_move.type == LastMove::Type::FlipCard) {
+        m_last_move.cards.at(0).set_upside_down(true);
+        if (on_undo_available)
+            on_undo_available(false);
+        invalidate_layout();
+        return;
+    }
+
+    if (m_last_move.from->type() == CardStack::Type::Play && m_mode == Mode::SingleCardDraw) {
+        auto& waste = stack(Waste);
+        if (!m_last_move.from->is_empty())
+            waste.push(m_last_move.from->pop());
+    }
+
+    for (auto& to_intersect : m_last_move.cards) {
+        mark_intersecting_stacks_dirty(to_intersect);
+        m_last_move.from->push(to_intersect);
+        m_last_move.to->pop();
+    }
+
+    if (m_last_move.from->type() == CardStack::Type::Stock) {
+        auto& waste = this->stack(Waste);
+        auto& play = this->stack(Play);
+        NonnullRefPtrVector<Card> cards_popped;
+        for (size_t i = 0; i < m_last_move.cards.size(); i++) {
+            if (!waste.is_empty()) {
+                auto card = waste.pop();
+                cards_popped.prepend(card);
+            }
+        }
+        for (auto& card : cards_popped) {
+            m_focused_cards.append(card);
+            play.push(move(card));
+        }
+    }
+
+    score_move(*m_last_move.from, *m_last_move.to, true);
+
+    m_last_move.from = nullptr;
+    m_last_move.to = nullptr;
```

Comment:
Hmm, so instead of resetting individual fields, you should be able to do:
```c++
m_last_move = {};
```

To reset all fields to their default values.

It might also make sense to add an `Invalid` state to `LastMove::Type`, and default-initialize the `LastMove::type` member to that value. Then, instead of checking `if (!m_last_move.from)` at the top of this function, you can check `if (m_last_move.type == LastMove::Type::Invalid)` to avoid the undo flip issue.

Reply:
Thanks for the tip, was not aware I could `= {};` the structure. I've also incorporated the `Invalid` suggestion, thanks for that.